SchemaMigration/COSchemaMigrationDriver.m

```diff
@@ -0,0 +1,428 @@
+	   in the model description repository. In that case we have no migration to do */
+	if (entity.owner != nil
+		&& entity.owner.version == item.packageVersion)
+	{
+		return [NSSet set];
+	}
+	
+	/* At this point we are doing a migration for some packages for sure. */
+
+	NSDictionary *versionsByDomain = [self versionsByDomainForItem: item];
+	
+	if (versionsByDomain == nil
+		|| versionsByDomain[@"org.etoile-project.CoreObject"] == nil
+		|| ![versionsByDomain[item.packageName] isEqual: @(item.packageVersion)])
+	{
+		// TODO: Test that we get this exception when needed
+		[NSException raise: NSInternalInconsistencyException
+					format: @"Item with entityName '%@' package '%@' version %d needs a migration, "
+							"but -versionsByDomainForItem: returned an incomplete "
+							"snapshot of the past version of the metamodel we need. "
+							"It returned: %@. \n"
+							"We require it to be a non-nil dictionary, have a "
+							"version set for the org.etoile-project.CoreObject "
+							"package, and include the same package/version as "
+							"the item being migrated. Probably, you forgot to "
+							"set COSchemaMigration.dependentSourceVersionsByDomain.",
+							item.entityName, item.packageName, (int)item.packageVersion, versionsByDomain];
+	}
+	
+	NSMutableSet *packagesToMigrate = [NSMutableSet new];
+	
+	for (NSString *package in [versionsByDomain allKeys])
+	{
+		ETPackageDescription *packageDesc = [_modelDescriptionRepository descriptionForName: package];
+		BOOL isDeletedPackage = (packageDesc == nil);
+		int64_t version = versionsByDomain[package] != nil
+			? (int64_t)[versionsByDomain[package] longLongValue]
+			: (int64_t)-1;
+
+		if (isDeletedPackage || version != (int64_t)packageDesc.version)
+		{
+			[packagesToMigrate addObject: package];
+		}
+	}
+	return packagesToMigrate;
+}
+
+- (BOOL)addItem: (COItem *)item
+{
+	ETAssert(_modelDescriptionRepository != nil);
+	NSSet *packagesToMigrate = [self packagesToMigrateForItem: item];
+
+	for (NSString *package in packagesToMigrate)
+	{
+		addObjectForKey(itemsToMigrate, item, package);
+	}
+	return ![packagesToMigrate isEmpty];
+}
+
+#pragma mark Triggering a Migration -
+
+- (NSArray *)prepareMigrationForItems: (NSArray *)storeItems
+{
+	NSMutableArray *upToDateItems = [NSMutableArray new];
+	
+	for (COItem *item in storeItems)
+	{
+		BOOL migrate = [self addItem: item];
+		
+		if (!migrate)
+		{
+			[upToDateItems addObject: item];
+		}
+	}
+	return upToDateItems;
+}
+
+- (NSArray *)migrateItems: (NSArray *)storeItems
+{
+	itemsToMigrate = [NSMutableDictionary new];
+	NSArray *upToDateItems = [self prepareMigrationForItems: storeItems];
+
+	for (NSString *packageName in [itemsToMigrate allKeys])
+	{
+		ETPackageDescription *package = [_modelDescriptionRepository descriptionForName: packageName];
+		ETAssert(package != nil);
+
+		// NOTE: Or -migrateItems:boundToPackageNamed:inRepository:
+		[self migrateItemsInDomain: packageName
```

Comment:
The NOTE should be removed, and the method -migrateItemsInDomain:toVersion: probably renamed to something like -migrateItemsBoundToPackageNamed:toVersion:

I was thinking of using BoundToPackageNamed: rather than InPackageNamed:, to make clearer the item can be processed as part of multiple packages in the migration process: the package owning it, plus packages on which this package depends on.